### PR TITLE
Support array submission with root workflow

### PR DESCRIPTION
**Pull request by Horneth**

### CODE REVIEW

File: engine/src/main/resources/swagger/cromwell.yaml [hunk 1]
```diff
@@ -53,6 +53,11 @@ paths:
           type: file
           in: formData
         - $ref: '#/parameters/workflowTypeParam'
+        - name: workflowRoot
+          description: The root object to be run. Only necessary for CWL submissions containing multiple objects (in an array).
+          required: false
+          type: string
+          in: formData
```
Comment (Horneth) [inline engine/src/main/resources/swagger/cromwell.yaml]: This is what I was talking about at stand up and what I mentioned yesterday @geoffjentry

File: cwl/src/main/scala/cwl/AddEmbeddedCwl.scala [hunk 2]
```diff
@@ -41,23 +43,25 @@ object AddEmbeddedCwl extends Poly1 {
...
     }
 
-  implicit def commandLineTool =
+  implicit def commandLineTool: Case.Aux[CommandLineTool, Map[String, Cwl] => Parse[Cwl]] =
     at[CommandLineTool] {
-      clt =>
-        Monad[Parse].pure(clt.asCwl)
+      clt => 
+        (_: Map[String, Cwl]) =>
+          Monad[Parse].pure(clt.asCwl)
```
Comment (mcovarr) [inline cwl/src/main/scala/cwl/AddEmbeddedCwl.scala]: I think you can `const` this and the below

File: cwl/src/main/scala/cwl/FlattenCwlFile.scala [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+      case None => s"Cannot find root object with id $root".invalidNelCheck
+    }
+  }
+  
+  private [cwl] object CwlFileRoot extends Poly1 {
+    implicit def fromCwl: Case.Aux[Cwl, Option[String] => Checked[Cwl]] = at[Cwl] { cwl => _ => cwl.validNelCheck }
+    implicit def fromCwlArray: Case.Aux[Array[Cwl], Option[String] => Checked[Cwl]] = at[Array[Cwl]] { cwls => 
+      _ match {
+        case Some(root) => findRoot(cwls, root)
+        case None => "The supplied file contains an array of Cwl objects but no root has been provided. Please provided the cwl root".invalidNelCheck
```
Comment (mcovarr) [inline cwl/src/main/scala/cwl/FlattenCwlFile.scala]: provided => provide

File: cwl/src/main/scala/cwl/command/ParentName.scala [hunk 4]
```diff
@@ -0,0 +1,10 @@
+package cwl.command
+
+object ParentName {
+  def empty: ParentName = ParentName(None) 
+  def apply(id: String): ParentName = ParentName(id.split("#").tail.headOption) 
+}
+
+case class ParentName(value: Option[String]) {
+  def stripParent(in: String) = value.map(v => in.stripPrefix(s"$v/")).getOrElse(in)
+}
```
Comment (mcovarr) [inline cwl/src/main/scala/cwl/command/ParentName.scala]: \n

File: cwl/src/main/scala/cwl/decoder/CoproductDecoder.scala [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+import shapeless.{:+:, CNil, Coproduct, Inl, Inr}
+import cats.syntax.show._
+
+/**
+  * Custom Coproduct decoder that keeps the last error message if all of the decoding attempts fail.
+  */
+trait CoproductDecoder {
+
+  implicit final val decodeCNil: Decoder[CNil] = new Decoder[CNil] {
+    def apply(c: HCursor): Decoder.Result[CNil] = Left(DecodingFailure("CNil", c.history))
```
Comment (mcovarr) [inline cwl/src/main/scala/cwl/decoder/CoproductDecoder.scala]: 😮

File: engine/src/main/scala/cromwell/webservice/PartialWorkflowSources.scala [hunk 6]
```diff
@@ -92,7 +94,7 @@ object PartialWorkflowSources {
         case (Some(_), Some(_)) => s"$WdlSourceKey and $WorkflowSourceKey can't both be supplied".invalidNel
         case (None, None) => s"$WorkflowSourceKey needs to be supplied".invalidNel
       }
-
+
```
Comment (mcovarr) [inline engine/src/main/scala/cromwell/webservice/PartialWorkflowSources.scala]: 👁

File: src/bin/travis/resources/conformance_expected_failures.txt [hunk 7]
```diff
@@ -72,7 +72,7 @@
 72
 73
 74
-75
+#75
 76
 77
 78
```
Comment (mcovarr) [inline src/bin/travis/resources/conformance_expected_failures.txt]: 🏆

**Pull request merged**